transport-native-epoll/src/main/java/io/netty/channel/epoll/AbstractEpollStreamChannel.java

```diff
@@ -672,4 +836,216 @@ public void run() {
+                        // - The user may want to to trigger another splice operation once the splicing was complete.
+                        config().setAutoRead(false);
+                    }
+                }
+
+                return len == 0;
+            } catch (Throwable cause) {
+                promise.setFailure(cause);
+                return true;
+            }
+        }
+    }
+
+    private final class SpliceOutTask {
+        private final AbstractEpollStreamChannel ch;
+        private final boolean autoRead;
+        private int len;
+
+        SpliceOutTask(AbstractEpollStreamChannel ch, int len, boolean autoRead) {
+            this.ch = ch;
+            this.len = len;
+            this.autoRead = autoRead;
+        }
+
+        public boolean spliceOut() throws Exception {
+            assert ch.eventLoop().inEventLoop();
+            try {
+                int splicedOut = Native.splice(ch.pipeIn, -1, ch.fd().intValue(), -1, len);
+                len -= splicedOut;
+                if (len == 0) {
+                    if (autoRead) {
+                        // AutoRead was used and we spliced everything so start reading again
+                        config().setAutoRead(true);
+                    }
+                    return true;
+                }
+                return false;
+            } catch (IOException e) {
+                if (autoRead) {
+                    // AutoRead was used and we spliced everything so start reading again
+                    config().setAutoRead(true);
+                }
+                throw e;
+            }
+        }
+    }
+
+    private final class SpliceFdTask extends SpliceInTask {
+        private final FileDescriptor fd;
+        private final ChannelPromise promise;
+        private int offset;
+
+        SpliceFdTask(FileDescriptor fd, int offset, int len, ChannelPromise promise) {
+            super(len, promise);
+            this.fd = fd;
+            this.promise = promise;
+            this.offset = offset;
+        }
+
+        @Override
+        public SpliceFdTask value() {
+            return this;
+        }
+
+        @Override
+        public boolean spliceIn(RecvByteBufAllocator.Handle handle) throws IOException {
+            assert eventLoop().inEventLoop();
+            if (len == 0) {
+                promise.setSuccess();
+                return true;
+            }
+
+            int pipeIn = -1;
+            int pipeOut = -1;
+            try {
+                long fds = Native.pipe();
+                pipeIn = (int) (fds >> 32);
+                pipeOut = (int) fds;
+
+                int splicedIn = spliceIn(pipeOut, handle);
+                if (splicedIn > 0) {
+                    // Integer.MAX_VALUE is a special value which will result in splice forever.
+                    if (len != Integer.MAX_VALUE) {
+                        len -= splicedIn;
+                    }
+                    while (splicedIn > 0) {
```

Comment:
consider using `do {} while();`. you already check `if (splicedIn > 0) {` above.